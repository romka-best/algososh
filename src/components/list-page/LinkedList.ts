export class LinkedListNode<T> {
    _value: T | null;
    next: LinkedListNode<T> | null;

    constructor(value: T | null, next: LinkedListNode<T> | null = null) {
        this._value = value;
        this.next = next;
    }
}

export default class LinkedList<T> {
    head: LinkedListNode<T> | null;
    tail: LinkedListNode<T> | null;

    constructor(defaultValues?: T[]) {
        this.head = null;
        this.tail = null;

        defaultValues?.forEach(value => this.append(value));
    }

    append(value: T): void {
        const node = new LinkedListNode(value);
        if (!this.head || !this.tail) {
            this.head = node;
            this.tail = node;
        }

        this.tail.next = node;
        this.tail = node;
    }

    prepend(value: T): void {
        const node = new LinkedListNode(value, this.head);
        this.head = node;

        if (!this.tail) {
            this.tail = node;
        }
    }

    addByIndex(index: number, value: T): void {
        if (index < 0 || index > this.toArray().length) return;

        let current = this.head;
        let currentIndex = 0;
        let previous;
        const node = new LinkedListNode(value);

        if (index === 0) {
            node.next = current;
            this.head = node;
        } else {
            while (currentIndex++ < index && current) {
                previous = current;
                current = current.next;
            }
            node.next = current;
            if (previous) {
                previous.next = node;
            }
        }
    }

    deleteByIndex(index: number): void {
        if (index > 0 && index > this.toArray().length) return;

        let current,
            previous,
            currentIndex = 0;
        current = this.head;
        previous = current;

        if (index === 0) {
            this.head = current ? current.next : null;
        } else {
            while (currentIndex < index) {
                currentIndex++;
                previous = current;
                current = current?.next;
            }
            if (previous) {
                previous.next = current ? current.next : null;
            }
        }
    }

    deleteHead(): LinkedListNode<T> | null {
        if (!this.head) {
            return null;
        }

        const deletedNode = this.head;

        if (this.head.next) {
            this.head = this.head.next;
        } else {
            this.head = null;
            this.tail = null;
        }

        return deletedNode;
    }

    deleteTail(): LinkedListNode<T> | null {
        if (!this.tail) {
            return null;
        }
        const deletedNode = this.tail;

        if (this.head === this.tail) {
            this.head = null;
            this.tail = null;

            return deletedNode;
        }

        let currentNode = this.head;
        while (currentNode?.next) {
            if (!currentNode.next.next) {
                this.tail = currentNode;
                currentNode.next = null;
            } else {
                currentNode = currentNode.next;
            }
        }

        return deletedNode;
    }

    toArray(): LinkedListNode<T>[] {
        const nodes = [];
        let currentNode = this.head;

        while (currentNode) {
            nodes.push(currentNode);
            currentNode = currentNode.next;
        }

        return nodes;
    }

    toInitialArray(): T[] {
        const nodes: T[] = [];
        let currentNode = this.head;

        while (currentNode) {
            nodes.push(currentNode._value as T);
            currentNode = currentNode.next;
        }

        return nodes;
    }
}
