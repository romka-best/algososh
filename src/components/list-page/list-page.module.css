.root {

}

.form {
    display: grid;
    grid-template-columns: repeat(5, 1fr);
    gap: 12px;
}

.input {
    max-width: 204px;
}

.button_big {
    grid-column: 2 / 4;
}

.button_big:first-of-type {
    grid-column: 2 / 4;
}

.button_big:last-of-type {
    grid-column: 4 / 6;
}

.circles {
    display: flex;
    justify-content: center;
    align-items: center;
    gap: 16px;
    flex-wrap: wrap;

    margin-top: 90px;
}
