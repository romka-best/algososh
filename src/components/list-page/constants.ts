import {OperationTypes} from "./list-page.types";

export const ADD_OPERATION_TYPE = [
    OperationTypes.AddByIndex,
    OperationTypes.AddHead,
    OperationTypes.AddTail,
];

export const DELETE_OPERATION_TYPE = [
    OperationTypes.DeleteByIndex,
    OperationTypes.DeleteHead,
    OperationTypes.DeleteTail,
];
