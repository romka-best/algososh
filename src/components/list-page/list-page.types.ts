import {LinkedListNode} from "./LinkedList";

export enum OperationTypes {
    AddHead = 'addHead',
    AddTail = 'addTail',
    DeleteHead = 'deleteHead',
    DeleteTail = 'deleteTail',
    AddByIndex = 'addByIndex',
    DeleteByIndex = 'deleteByIndex',
}

export interface Step<T> {
    list: LinkedListNode<T>[];
    index?: number,
    value?: T | null,
}
