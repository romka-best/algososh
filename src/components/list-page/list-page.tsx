import React from "react";

import {SolutionLayout} from "../ui/solution-layout/solution-layout";
import {Input} from "../ui/input/input";
import {Button} from "../ui/button/button";
import {Circle} from "../ui/circle/circle";
import {ArrowIcon} from "../ui/icons/arrow-icon";

import {useForm} from "../../hooks/useForm";

import LinkedList from "./LinkedList";

import {
    addByIndex,
    addHead,
    addTail, deleteByIndex,
    deleteHead, deleteTail,
    getLetterElementHead,
    getLetterElementTail,
    getLetterState
} from "./utils";

import {DELAY_IN_MS} from "../../constants/delays";

import {OperationTypes, Step} from "./list-page.types";

import styles from "./list-page.module.css";

const DEFAULT_VALUE = ["0", "34", "8", "1"]

export const ListPage: React.FC = () => {
    const [isLoading, setIsLoading] = React.useState<boolean>(false);
    // const [currentInputValue, setCurrentInputValue] = React.useState<string>("");
    // const [currentInputIndex, setCurrentInputIndex] = React.useState<string>("");

    const {values, handleChange, setValues} = useForm({
        value: '',
        index: '',
    });

    const list = React.useRef(new LinkedList(DEFAULT_VALUE));
    const intervalId = React.useRef<NodeJS.Timeout>();
    const [steps, setSteps] = React.useState<Step<string>[]>([{list: list.current.toArray()}]);
    const [currentStep, setCurrentStep] = React.useState<number>(0);
    const [currentOperation, setCurrentOperation] = React.useState<OperationTypes | null>(null);

    React.useEffect(() => {
        if (!currentOperation) return;

        setIsLoading(true);

        let steps: Step<string>[] = [];

        switch (currentOperation) {
            case OperationTypes.AddHead:
                steps = addHead<string>(values.value, list.current);
                break;
            case OperationTypes.AddTail:
                steps = addTail<string>(values.value, list.current);
                break;
            case OperationTypes.AddByIndex:
                steps = addByIndex<string>(values.value, Number(values.index), list.current);
                break;
            case OperationTypes.DeleteHead:
                steps = deleteHead<string>(list.current);
                break;
            case OperationTypes.DeleteTail:
                steps = deleteTail<string>(list.current);
                break;
            case OperationTypes.DeleteByIndex:
                steps = deleteByIndex<string>(Number(values.index), list.current);
                break;
        }

        if (steps.length > 1) {
            setSteps(steps);
            setCurrentStep(0);

            intervalId.current = setInterval(() => {
                setCurrentStep((currentStep) => {
                    if (currentStep === steps.length - 1 && intervalId.current) {
                        clearInterval(intervalId.current);
                        setCurrentOperation(null);
                        setIsLoading(false);
                        setSteps([steps[steps.length - 1]]);
                        setValues({
                            value: '',
                            index: '',
                        })

                        return 0;
                    }
                    return currentStep + 1;
                })
            }, DELAY_IN_MS);
        }
    }, [currentOperation, values]);

    return (
        <SolutionLayout title="Связный список">
            <div className={styles.root}>
                <form className={styles.form} onSubmit={(event: React.FormEvent<HTMLFormElement>) => {
                    event.preventDefault();
                }}>
                    <Input
                        extraClass={styles.input}
                        placeholder="Введите значение"
                        name="value"
                        isLimitText={true}
                        maxLength={4}
                        disabled={isLoading}
                        value={values.value}
                        onChange={handleChange}
                    />
                    <Button
                        text="Добавить в head"
                        disabled={isLoading || !values.value.length}
                        isLoader={isLoading && currentOperation === OperationTypes.AddHead}
                        onClick={() => {
                            setCurrentOperation(OperationTypes.AddHead);
                        }}
                        type="submit"
                    />
                    <Button
                        text="Добавить в tail"
                        disabled={isLoading || !values.value.length}
                        isLoader={isLoading && currentOperation === OperationTypes.AddTail}
                        onClick={() => {
                            setCurrentOperation(OperationTypes.AddTail);
                        }}
                        type="submit"
                    />
                    <Button
                        text="Удалить из head"
                        disabled={isLoading || !steps[currentStep].list.length}
                        isLoader={isLoading && currentOperation === OperationTypes.DeleteHead}
                        onClick={() => {
                            setCurrentOperation(OperationTypes.DeleteHead);
                        }}
                        type="submit"
                    />
                    <Button
                        text="Удалить из tail"
                        disabled={isLoading || !steps[currentStep].list.length}
                        isLoader={isLoading && currentOperation === OperationTypes.DeleteTail}
                        onClick={() => {
                            setCurrentOperation(OperationTypes.DeleteTail);
                        }}
                        type="submit"
                    />
                    <Input
                        extraClass={styles.input}
                        placeholder="Введите индекс"
                        name="index"
                        type="number"
                        disabled={isLoading}
                        value={values.index}
                        onChange={handleChange}
                    />
                    <Button
                        extraClass={styles.button_big}
                        text="Добавить по индексу"
                        disabled={isLoading ||
                            !values.index.length ||
                            !values.value.length ||
                            Number.isNaN(Number(values.index)) ||
                            Number(values.index) < 0 ||
                            Number(values.index) >= steps[currentStep].list.length
                        }
                        isLoader={isLoading && currentOperation === OperationTypes.AddByIndex}
                        onClick={() => {
                            setCurrentOperation(OperationTypes.AddByIndex);
                        }}
                        type="submit"
                    />
                    <Button
                        extraClass={styles.button_big}
                        text="Удалить по индексу"
                        disabled={isLoading ||
                            !values.index.length ||
                            !steps[currentStep].list.length ||
                            Number.isNaN(Number(values.index)) ||
                            Number(values.index) < 0 ||
                            Number(values.index) >= steps[currentStep].list.length
                        }
                        isLoader={isLoading && currentOperation === OperationTypes.DeleteByIndex}
                        onClick={() => {
                            setCurrentOperation(OperationTypes.DeleteByIndex);
                        }}
                        type="submit"
                    />
                </form>
                <div className={styles.circles}>
                    {
                        steps[currentStep].list?.map((node, index) => (
                            <React.Fragment key={index}>
                                <Circle
                                    index={index}
                                    letter={node._value ? node._value : ""}
                                    tail={getLetterElementTail(steps[currentStep], index, currentOperation, steps[currentStep].list.length)}
                                    head={getLetterElementHead(steps[currentStep], index, currentOperation)}
                                    state={getLetterState(index, steps[currentStep], currentOperation)}
                                />
                                {steps[currentStep].list.length - 1 !== index && <ArrowIcon/>}
                            </React.Fragment>
                        ))
                    }
                </div>
            </div>
        </SolutionLayout>
    );
};
