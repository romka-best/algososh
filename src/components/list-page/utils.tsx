import React from "react";

import {Circle} from "../ui/circle/circle";

import LinkedList from "./LinkedList";

import {HEAD, TAIL} from "../../constants/element-captions";
import {ADD_OPERATION_TYPE, DELETE_OPERATION_TYPE} from "./constants";

import {ElementStates} from "../../types/element-states";
import {OperationTypes, Step} from "./list-page.types";

export function addHead<T>(value: T, list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    steps.push({
        index: 0,
        value,
        list: list.toArray(),
    });

    list.prepend(value);

    steps.push({
        index: 0,
        list: list.toArray(),
    });

    steps.push({
        list: list.toArray(),
    });

    return steps;
}

export function addTail<T>(value: T, list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    let currentList = list.toArray();
    steps.push({
        index: currentList.length - 1,
        value,
        list: currentList,
    });

    list.append(value);

    currentList = list.toArray();
    steps.push({
        index: currentList.length - 1,
        list: currentList,
    });

    steps.push({
        list: currentList,
    });

    return steps;
}

export function addByIndex<T>(value: T, index: number, list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];
    const currentList = list.toArray();

    let currentIndex = 0;
    while (currentIndex !== index) {
        steps.push({
            index: currentIndex,
            value,
            list: currentList,
        });
        currentIndex++;
    }

    list.addByIndex(index, value);

    steps.push({
        index: currentIndex,
        list: list.toArray(),
    });

    steps.push({
        list: list.toArray(),
    });

    return steps;
}

export function deleteHead<T>(list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    const currentList = list.toArray();
    const value = list.head?._value;
    currentList[0]._value = null;
    steps.push({
        index: 0,
        value,
        list: currentList,
    });

    list.deleteHead();

    steps.push({
        index: 0,
        list: list.toArray(),
    });

    steps.push({
        list: list.toArray(),
    });

    return steps;
}

export function deleteTail<T>(list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    let currentList = list.toArray();
    const value = list.tail?._value;
    currentList[currentList.length - 1]._value = null;
    steps.push({
        index: currentList.length - 1,
        value,
        list: currentList,
    });

    list.deleteTail();

    currentList = list.toArray();

    steps.push({
        index: currentList.length - 1,
        list: currentList,
    });

    steps.push({
        list: currentList,
    });

    return steps;
}

export function deleteByIndex<T>(index: number, list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    let currentList = list.toArray();

    let currentIndex = 0;
    while (currentIndex !== index) {
        steps.push({
            index: currentIndex,
            list: currentList,
        });
        currentIndex++;
    }

    const prevList = new LinkedList(list.toInitialArray()).toArray();
    const value = prevList[currentIndex]._value;
    prevList[currentIndex]._value = null;
    steps.push({
        index: currentIndex,
        value,
        list: prevList,
    });

    list.deleteByIndex(index);

    currentList = list.toArray();

    steps.push({
        index: currentIndex,
        list: currentList,
    });

    steps.push({
        list: currentList,
    });

    return steps;
}

export function getLetterState<T>(
    index: number,
    currentStep: Step<T>,
    operationType: OperationTypes | null,
): ElementStates {
    if (!operationType || currentStep.index === undefined) {
        return ElementStates.Default;
    }

    if ((currentStep.value && ((operationType === "addHead" || operationType === "addByIndex") &&
            index < currentStep.index) || (operationType === "addTail" && index > currentStep.index)) ||
        (operationType === "deleteByIndex" && index <= currentStep.index)
    ) {
        return ElementStates.Changing;
    }

    if (!currentStep.value && currentStep.index === index) {
        return ElementStates.Modified;
    }

    return ElementStates.Default;
}

export function getLetterElementHead(step: Step<string>, index: number, operation: OperationTypes | null): string | React.ReactElement {
    if (operation &&
        step.value &&
        step.index === index &&
        ADD_OPERATION_TYPE.includes(operation)) {
        return (
            <Circle
                letter={step.value}
                state={ElementStates.Changing}
                isSmall
            />
        )
    }

    if (!index) {
        return HEAD;
    }

    return "";
}

export function getLetterElementTail(step: Step<string>,
                                     index: number,
                                     operation: OperationTypes | null,
                                     length: number): string | React.ReactElement {
    if (operation &&
        step.value &&
        step.index === index &&
        DELETE_OPERATION_TYPE.includes(operation)) {
        return (
            <Circle
                letter={step.value}
                state={ElementStates.Changing}
                isSmall
            />
        )
    }

    if (index === length - 1) {
        return TAIL;
    }

    return "";
}
