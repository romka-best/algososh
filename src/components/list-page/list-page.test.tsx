import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import { ListPage } from './list-page';

describe('ListPage', () => {
    describe('Отображение', () => {
        it('Начальное состяние', () => {
            const page = renderer
                .create(
                    <BrowserRouter>
                        <ListPage/>
                    </BrowserRouter>
                )
                .toJSON();

            expect(page).toMatchSnapshot();
        });
    });
});
