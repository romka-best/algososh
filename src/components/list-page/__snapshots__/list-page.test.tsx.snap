// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPage Отображение Начальное состяние 1`] = `
<main
  className="content"
>
  <div
    className="titleBox"
  >
    <h1
      className="text text_type_h2 text_color_h1 title"
    >
      МБОУ АЛГОСОШ
    </h1>
    <span
      className="text text_type_fibonacci text_color_secondary subtitle"
    >
      им. Фибоначчи
    </span>
  </div>
  <div
    className="contentCard"
  >
    <a
      className="link"
      href="/"
      onClick={[Function]}
    >
      <button
        className="button returnButton"
        type="button"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M12.7073 4.29289C13.0978 4.68342 13.0978 5.31658 12.7073 5.70711L7.41436 11H19.0002C19.5524 11 20.0002 11.4477 20.0002 12C20.0002 12.5523 19.5524 13 19.0002 13H7.41436L12.7073 18.2929C13.0978 18.6834 13.0978 19.3166 12.7073 19.7071C12.3167 20.0976 11.6836 20.0976 11.293 19.7071L3.58594 12L11.293 4.29289C11.6836 3.90237 12.3167 3.90237 12.7073 4.29289Z"
            fill="#0032FF"
            fillRule="evenodd"
          />
        </svg>
        <p
          className="text text_type_button text_color_link ml-4"
        >
          К оглавлению
        </p>
      </button>
    </a>
    <h3
      className="text text_type_h3 text_color_h3 cardTitle"
    >
      Связный список
    </h3>
    <div
      className="root"
    >
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="content input"
        >
          <input
            className="input text text_type_input text_color_input"
            disabled={false}
            maxLength={4}
            name="value"
            onChange={[Function]}
            placeholder="Введите значение"
            type="text"
            value=""
          />
          <span
            className="text text_type_input-lim text_color_input mt-2 ml-8 limit"
          >
            Максимум — 4 символа
          </span>
        </div>
        <button
          className="text text_type_button text_color_primary button"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          <p
            className="text"
          >
            Добавить в head
          </p>
        </button>
        <button
          className="text text_type_button text_color_primary button"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          <p
            className="text"
          >
            Добавить в tail
          </p>
        </button>
        <button
          className="text text_type_button text_color_primary button"
          disabled={false}
          onClick={[Function]}
          type="submit"
        >
          <p
            className="text"
          >
            Удалить из head
          </p>
        </button>
        <button
          className="text text_type_button text_color_primary button"
          disabled={false}
          onClick={[Function]}
          type="submit"
        >
          <p
            className="text"
          >
            Удалить из tail
          </p>
        </button>
        <div
          className="content input"
        >
          <input
            className="input text text_type_input text_color_input"
            disabled={false}
            name="index"
            onChange={[Function]}
            placeholder="Введите индекс"
            type="number"
            value=""
          />
        </div>
        <button
          className="text text_type_button text_color_primary button button_big"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          <p
            className="text"
          >
            Добавить по индексу
          </p>
        </button>
        <button
          className="text text_type_button text_color_primary button button_big"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          <p
            className="text"
          >
            Удалить по индексу
          </p>
        </button>
      </form>
      <div
        className="circles"
      >
        <div
          className="content"
        >
          <div
            className="text text_type_input text_color_input mb-4 absolute head string"
          >
            head
          </div>
          <div
            className="circle default"
          >
            <p
              className="text text_type_circle text_color_input letter"
            >
              0
            </p>
          </div>
          <p
            className="text text_type_input text_color_input mt-4 absolute index"
          >
            0
          </p>
          <div
            className="text text_type_input text_color_input mt-4 absolute tail tail60 string"
          >
            
          </div>
        </div>
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M9.29289 4.29289C9.68342 3.90237 10.3166 3.90237 10.7071 4.29289L18.4142 12L10.7071 19.7071C10.3166 20.0976 9.68342 20.0976 9.29289 19.7071C8.90237 19.3166 8.90237 18.6834 9.29289 18.2929L15.5858 12L9.29289 5.70711C8.90237 5.31658 8.90237 4.68342 9.29289 4.29289Z"
            fill="#0032FF"
            fillRule="evenodd"
          />
        </svg>
        <div
          className="content"
        >
          <div
            className="text text_type_input text_color_input mb-4 absolute head string"
          >
            
          </div>
          <div
            className="circle default"
          >
            <p
              className="text text_type_circle text_color_input letter"
            >
              34
            </p>
          </div>
          <p
            className="text text_type_input text_color_input mt-4 absolute index"
          >
            1
          </p>
          <div
            className="text text_type_input text_color_input mt-4 absolute tail tail60 string"
          >
            
          </div>
        </div>
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M9.29289 4.29289C9.68342 3.90237 10.3166 3.90237 10.7071 4.29289L18.4142 12L10.7071 19.7071C10.3166 20.0976 9.68342 20.0976 9.29289 19.7071C8.90237 19.3166 8.90237 18.6834 9.29289 18.2929L15.5858 12L9.29289 5.70711C8.90237 5.31658 8.90237 4.68342 9.29289 4.29289Z"
            fill="#0032FF"
            fillRule="evenodd"
          />
        </svg>
        <div
          className="content"
        >
          <div
            className="text text_type_input text_color_input mb-4 absolute head string"
          >
            
          </div>
          <div
            className="circle default"
          >
            <p
              className="text text_type_circle text_color_input letter"
            >
              8
            </p>
          </div>
          <p
            className="text text_type_input text_color_input mt-4 absolute index"
          >
            2
          </p>
          <div
            className="text text_type_input text_color_input mt-4 absolute tail tail60 string"
          >
            
          </div>
        </div>
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M9.29289 4.29289C9.68342 3.90237 10.3166 3.90237 10.7071 4.29289L18.4142 12L10.7071 19.7071C10.3166 20.0976 9.68342 20.0976 9.29289 19.7071C8.90237 19.3166 8.90237 18.6834 9.29289 18.2929L15.5858 12L9.29289 5.70711C8.90237 5.31658 8.90237 4.68342 9.29289 4.29289Z"
            fill="#0032FF"
            fillRule="evenodd"
          />
        </svg>
        <div
          className="content"
        >
          <div
            className="text text_type_input text_color_input mb-4 absolute head string"
          >
            
          </div>
          <div
            className="circle default"
          >
            <p
              className="text text_type_circle text_color_input letter"
            >
              1
            </p>
          </div>
          <p
            className="text text_type_input text_color_input mt-4 absolute index"
          >
            3
          </p>
          <div
            className="text text_type_input text_color_input mt-4 absolute tail tail60 string"
          >
            tail
          </div>
        </div>
      </div>
    </div>
  </div>
  <p
    className="text text_type_column text_color_input mt-14 copyright"
  >
    © Сделано в Практикуме Романом Даниловым.
  </p>
</main>
`;
